added resolve pagerduty incidents docs (#1324)
port-labs/port-docs: docs/guides-and-tutorials/resolve-pagerduty-incident.md

---
sidebar_position: 12
tags:
  - Automations
  - PagerDuty
  - Slack
  - Incident
  - GitHub
---

import PortTooltip from "/src/components/tooltip/tooltip.jsx"
import PortApiRegionTip from "/docs/generalTemplates/_port_region_parameter_explanation_template.md"

# Resolve Pagerduty incidents

## Overview

Solving incidents efficiently is a crucial part of any production-ready environment. When managing an incident, there are a few base concepts which are important to keep:

> - **Real time notifications** - When an incident has been created, either by an alert or manually, it is important that a push notification will be sent the the relevant owners and stakeholders as soon as possible. This can be in the form of a Slack message, email or any other form of communication.

Reviewer: This also happens in the previous guide, so please fix it there too

permissions:
  contents: read
  issues: write 

jobs:
  resolve_incident:
    runs-on: ubuntu-latest
    env:
      PD_INCIDENT_ID: ${{ fromJson(inputs.port_payload).entity.identifier }}
      PD_INCIDENT_URL: ${{ fromJson(inputs.port_payload).entity.properties.url }}
      PD_INCIDENT_TITLE: ${{ fromJson(inputs.port_payload).entity.title }}
      PORT_INCIDENT_URL: https://app.getport.io/pagerdutyIncidentEntity?identifier=${{ fromJson(inputs.port_payload).entity.identifier }}
      PORT_RUN_ID: ${{fromJson(inputs.port_payload).run_id}}
    steps:
      - name: Log Executing Request to Resolve Incident
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{ env.PORT_RUN_ID }}
          logMessage: "Resloving Pagerduty incident '${{ env.PD_INCIDENT_ID }}'..."

      - name: Request to Resolve Incident
        id: resolve_incident
        uses: fjogeleit/http-request-action@v1
        with:
          url: 'https://api.pagerduty.com/incidents'
          method: 'PUT'

>           customHeaders: '{"Content-Type": "application/json", "Accept": "application/vnd.pagerduty+json;version=2", "Authorization": "Token token=${{ secrets.PAGERDUTY_API_KEY }}", "From": "aknmhlxqagykzpwxxs@cazlq.com"}'

Author: Will need to match a PagerDuty user that can resolve the issue, I'll figure it out

      - name: Send Slack Message
        uses: archive/github-actions-slack@v2.9.0
        id: send-message
        with:
          slack-function: send-message
          slack-bot-user-oauth-access-token: ${{ secrets.BOT_USER_OAUTH_TOKEN }}
          slack-channel: ${{fromJson(inputs.port_payload).slack_channel_id}}
          slack-text: | 
            🚀 Incident was resolved 🚀
            View incident :point_right::skin-tone-4: <${{ env.PORT_INCIDENT_URL }}|here>!
            Good job everyone, thank you for your help 💪🏻

      - name: Finished handling resolution log
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{ env.PORT_RUN_ID }}
          logMessage: |
            Incident '${{ env.PD_INCIDENT_ID }}' resolved successfully 💪🏻
```

<PortApiRegionTip/>

</details>

We also need to create the following secrets in our GitHub repository:
- `PORT_CLIENT_ID` - Your Port client ID.
- `PORT_CLIENT_SECRET` - Your Port client secret.

> - `BOT_USER_OAUTH_TOKEN` - The Slack app bot token.

Reviewer: Maybe add in brackets that they get it from the previous guide

Author: Modified, let me know what you think

## Port Action Configuration
Let's create the Port Self-service action:
1. Navigate to the `Self-service` tab in your Port organization.
2. Press `New action`.
3. Choose the `Edit JSON` option.
4. Create the action with the following JSON definition:
<details>

  <summary><b>Port Action: Resolve Incident (click to expand)</b></summary>

>    :::tip

Reviewer: ```suggestion
   :::tip action placeholders
```

   Replace the following placeholders to match your environment: 
- `<GITHUB-ORG>` - your GitHub organization or user name.
- `<GITHUB-REPO-NAME>` - your GitHub repository name.
:::

```json showLineNumbers
{
  "identifier": "pagerduty_resolve_incident",
  "title": "Resolve Incident",
  "icon": "pagerduty",

>   "description": "Resolve incident in pagerduty",

Reviewer: ```suggestion
  "description": "Resolve incident in PagerDuty",
```

  "trigger": {
    "type": "self-service",
    "operation": "DAY-2",
    "userInputs": {
      "properties": {},
      "required": [],
      "order": []
    },
    "blueprintIdentifier": "pagerdutyIncident"
  },
  "invocationMethod": {
    "type": "GITHUB",
    "org": "<GITHUB-ORG>",
    "repo": "<GITHUB-REPO-NAME>",
    "workflow": "resolve-incident.yaml",
    "workflowInputs": {
      "{{if (.inputs | has(\"ref\")) then \"ref\" else null end}}": "{{.inputs.\"ref\"}}",
      "{{if (.inputs | has(\"from\")) then \"from\" else null end}}": "{{.inputs.\"from\"}}",
      "port_payload": {
        "blueprint": "{{.action.blueprint}}",
        "slack_channel_id": "{{.entity.properties.slack_channel | split(\"=\") | .[-1]}}",
        "gh_issue_id": "{{.entity.relations.githubIssue | split(\"-\") | .[-1]}}",
        "entity": "{{.entity}}",
        "run_id": "{{.run.id}}"
      }
    },
    "reportWorkflowStatus": true
  },
  "requiredApproval": false,
  "publish": true
}
```

</details>

### Testing the Action
Let's test our action flow:
1. Head over to your Port organization.
2. Navigate to the `Self-service` tab.

> 3. Find the new action `ACTION_NAME` and press `Execute`.

Reviewer: ACTION_NAME is not right

4. Fill the Pagerduty Incident you would like to resolve and press `Execute`.

> 5. Go to your Slack workspace and make sure the incident channel was archived.

Reviewer: The workflow does not perform any action to archive it from what I can tell, we should probably change this to say - Go to your Slack workspace and make sure that the incident resolution was sent to the channel

Author: fixed

6. Head over to your Github repository and make sure the issue was closed.
7. Navigate to your [PagerDuty Incidents](https://app.getport.io/pagerdutyIncidents) entities page. Make sure the incident's status changed to `resolved`.

## Summary 

> While in the previous guide we used Port as the automation orchestrator to manage the incident flow, this action will make it easier to resolve Pagerduty incidents including cleaning up all the additional resources we created for the incident in the automated process.

Reviewer: ```suggestion
While in the previous guide we used Port as the automation orchestrator to manage the incident flow, this action will make it easier to resolve PagerDuty incidents including cleaning up all the additional resources we created for the incident in the automated process.
```

## Next Steps

> - Add an Automation to resolve incidents that were not resolved by our newly created Port action, in order to avoid stale Slack channels and Github issues.

Reviewer: ```suggestion
- Add an Automation that will enhance the action in this guide, and will automatically trigger the incident resolution workflow, to perform automatic cleanup of incidents.
```

While it is important to efficiently manage an incident as it is being addressed, it is also just as important to efficiently summarize the incident and perform cleanup. In this guide, we will be using Port's [Self-Service Actions](https://docs.getport.io/actions-and-automations/create-self-service-experiences/) capabilities to efficiently resolve and cleanup the resources related to the PagerDuty incident.

## Prerequisites
- Complete the [Automating incident management](https://docs.getport.io/guides-and-tutorials/create-slack-channel-for-reported-incident) guide.
- User email of a member in your PagerDuty account.

## Data model setup
For this guide, we will be using the same data model as in the [Automating incident management](https://docs.getport.io/guides-and-tutorials/create-slack-channel-for-reported-incident) guide.

### Action backend
As a backend for the Port Action, create a GitHub Workflow in your repository.

Create a workflow file with the following content:

<details>
    <summary>`Resolve incident` GitHub workflow YAML</summary>

This workflow is responsible for resolving an incident, notifying the Slack channel and closing the GitHub issue.

> :::tip

Reviewer: ```suggestion
:::tip Update placeholders
```

>    Replace the `<PAEGRDUTY_USER_EMAIL>` placeholder to the user email from the [Prerequisites](#prerequisites) section. 

Reviewer: ```suggestion
   Replace the `<PAEGRDUTY_USER_EMAIL>` placeholder with the user email from the [prerequisites](#prerequisites) section. 
```

          runId: ${{ env.PORT_RUN_ID }}
          logMessage: |
            Github issue closed successfully ✅

            Updating the Slack channel that the incident was resolved...

      - name: Send Slack Message
        uses: archive/github-actions-slack@v2.9.0
        id: send-message
        with:
          slack-function: send-message
          slack-bot-user-oauth-access-token: ${{ secrets.BOT_USER_OAUTH_TOKEN }}
          slack-channel: ${{fromJson(inputs.port_payload).slack_channel_id}}
          slack-text: | 
            🚀 Incident was resolved 🚀
            View incident :point_right::skin-tone-4: <${{ env.PORT_INCIDENT_URL }}|here>!
            Good job everyone, thank you for your help 💪🏻

      - name: Finished handling resolution log
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{ env.PORT_RUN_ID }}
          logMessage: |
            Incident '${{ env.PD_INCIDENT_ID }}' resolved successfully 💪🏻
```

<PortApiRegionTip/>

</details>

> Make sure you created the following secrets in your GitHub repository (Should have done it in the [automation guide](https://docs.getport.io/guides-and-tutorials/create-slack-channel-for-reported-incident)):

Reviewer: ```suggestion
Make sure you created the following secrets in your GitHub repository (If you followed the [incident automation guide](https://docs.getport.io/guides-and-tutorials/create-slack-channel-for-reported-incident) then they should all already be in place):
```
